### PR TITLE
Add maximum file size for bundling wasm assets

### LINKED ISSUE #112: Gracefully handle R packages too large for GitHub Pages
Some ideas:

* Provide `shinylive::export()` with a blocklist of packages to ignore.

* Reject bundling packages over a certain configurable size, defaulting to 100MB.

### PR DESCRIPTION
**Pull request by georgestagg**
Closes #112.

### CODE REVIEW

File: R/packages.R [hunk 1]
```diff
@@ -1,3 +1,10 @@
+SHINYLIVE_DEFAULT_MAX_FILESIZE <- "100MB"
+
+# Default maximum filesize for asset bundling
+default_max_filesize <- function() {
+  Sys.getenv("SHINYLIVE_DEFAULT_MAX_FILESIZE", SHINYLIVE_DEFAULT_MAX_FILESIZE)
+}
```
Comment (schloerke) [inline R/packages.R:6]: Feels weird to have two locations that are reading in the sys envar

```suggestion
# Sys env maximum filesize for asset bundling
sys_env_max_filesize <- function() {
  Sys.getenv("SHINYLIVE_DEFAULT_MAX_FILESIZE", NULL)
}
```

File: R/packages.R [hunk 2]
```diff
@@ -178,7 +185,14 @@ env_download_wasm_core_packages <- function() {
   strsplit(pkgs, "\\s*[ ,\n]\\s*")[[1]]
 }
 
-download_wasm_packages <- function(appdir, destdir, package_cache) {
+download_wasm_packages <- function(appdir, destdir, package_cache, max_filesize) {
+  max_filesize_missing <- Sys.getenv("SHINYLIVE_DEFAULT_MAX_FILESIZE") == "" && is.null(max_filesize)
```
Comment (schloerke) [inline R/packages.R:189]:
```suggestion
  max_filesize_missing <- is.null(sys_env_max_filesize()) && is.null(max_filesize)
```

File: R/packages.R [hunk 3]
```diff
@@ -178,7 +185,14 @@ env_download_wasm_core_packages <- function() {
...
+  max_filesize_cli_fn <- if (max_filesize_missing) cli::cli_warn else cli::cli_abort
+
+  max_filesize <- max_filesize %||% default_max_filesize()
```
Comment (schloerke) [inline R/packages.R:192]:
```suggestion
  max_filesize <- max_filesize %||% sys_env_max_filesize() %||% SHINYLIVE_DEFAULT_MAX_FILESIZE
```

File: R/packages.R [hunk 4]
```diff
@@ -178,7 +185,14 @@ env_download_wasm_core_packages <- function() {
...
+  max_filesize <- if (is.na(max_filesize) || (max_filesize < 0)) Inf else max_filesize
+  max_filesize <- fs::fs_bytes(max_filesize)
```
Comment (schloerke) [inline R/packages.R:194]: Safe guard parsed value

```suggestion
  max_filesize_val <- max_filesize
  max_filesize <- fs::fs_bytes(max_filesize)
  if (is.na(max_filesize)) {
    cli::cli_warn(c(
      "!" = "Could not parse `max_filesize` value: {.code   max_filesize_val }",
      "i" = "Setting to {.code SHINYLIVE_DEFAULT_MAX_FILESIZE }"
    ))
    max_filesize <- fs::fs_bytes(SHINYLIVE_DEFAULT_MAX_FILESIZE)
  }
```

**Review by schloerke — APPROVED** [review]
Untested suggestions.
 
Thank you!

File: R/packages.R (hunk 1 above)
Comment (georgestagg) [reply R/packages.R]: It looks like `Sys.getenv()` doesn't like `NULL` as its fallback value, it wants a character vector there.

I've gone with:
```r
sys_env_max_filesize <- function() {
  max_fs_env <- Sys.getenv("SHINYLIVE_DEFAULT_MAX_FILESIZE")
  if (max_fs_env == "") NULL else max_fs_env
}
```

File: R/packages.R (hunk 4 above)
Comment (georgestagg) [reply R/packages.R:196]: Minor tweak made here to use e.g. `{.code {max_filesize_val}}` so that the values are printed by cli.

**Comment by georgestagg** [comment]
Thanks, suggestions committed!

**Pull request merged**